larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MatchInfo::MatchInfo() :
+    MatchInfo(QualityCuts())
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MatchInfo::MatchInfo(const QualityCuts &qualityCuts) :
+    m_qualityCuts{qualityCuts}
+{
+}
+
+void LArHierarchyHelper::MatchInfo::Match(const MCHierarchy &mcHierarchy, const RecoHierarchy &recoHierarchy)
+{
+    MCHierarchy::NodeVector mcNodes;
+    mcHierarchy.GetFlattenedNodes(mcNodes);
+    RecoHierarchy::NodeVector recoNodes;
+    recoHierarchy.GetFlattenedNodes(recoNodes);
+
+    std::sort(mcNodes.begin(), mcNodes.end(), [](const MCHierarchy::Node *lhs, const MCHierarchy::Node *rhs)
+        {
+            return lhs->GetCaloHits().size() > rhs->GetCaloHits().size();
+        });
+    std::sort(recoNodes.begin(), recoNodes.end(), [](const RecoHierarchy::Node *lhs, const RecoHierarchy::Node *rhs)
+        {
+            return lhs->GetCaloHits().size() > rhs->GetCaloHits().size();
+        });
+
+    std::map<const MCHierarchy::Node *, MCMatches> mcToMatchMap;
+    for (const RecoHierarchy::Node *pRecoNode : recoNodes)
+    {
+        const CaloHitList &recoHits{pRecoNode->GetCaloHits()};
+        const MCHierarchy::Node *pBestNode{nullptr};
+        size_t bestSharedHits{0};
+        for (const MCHierarchy::Node *pMCNode : mcNodes)
+        {
+            if (!pMCNode->IsReconstructable())
+                continue;
+            const CaloHitList &mcHits{pMCNode->GetCaloHits()};
+            CaloHitVector intersection;
+            std::set_intersection(mcHits.begin(), mcHits.end(), recoHits.begin(), recoHits.end(), std::back_inserter(intersection));
+
+            if (!intersection.empty())
+            {
+                const size_t sharedHits{intersection.size()};
+                if (sharedHits > bestSharedHits)
+                {
+                    bestSharedHits = sharedHits;
+                    pBestNode = pMCNode;
+                }
+            }
+        }
+        if (pBestNode)
+        {
+            auto iter{mcToMatchMap.find(pBestNode)};
+            if (iter != mcToMatchMap.end())
+            {
+                MCMatches &match(iter->second);
+                match.AddRecoMatch(pRecoNode, static_cast<int>(bestSharedHits));
+            }
+            else
+            {
+                MCMatches match(pBestNode);
+                match.AddRecoMatch(pRecoNode, static_cast<int>(bestSharedHits));
+                mcToMatchMap.insert(std::make_pair(pBestNode, match));
+            }
+        }
+        else
+        {
+            m_unmatchedReco.emplace_back(pRecoNode);
+        }
+    }
+
+    for (auto [ pMCNode, matches ] : mcToMatchMap)
+    {   (void)pMCNode;  // gcc 7 requirement
```

Comment:
Is there a reason why this is on the same line as the {?

Reply:
Just a personal preference for these sort of technical requirements. It has no relevance to the logic of the code and it stands out here as something that can be removed as and when we drop support for legacy versions of gcc.